File: java-pi-pantry/src/main/java/com/langrsoft/pi/pantry/Pantry.java
package com.langrsoft.pi.pantry;

import java.time.Clock;
import java.time.LocalDate;
import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Pantry {
    private Map<String, List<Item>> items = new HashMap<>();
    private Clock clock = Clock.systemDefaultZone();
    private static final int CAPACITY = 512;

    public void purchase(Item item) {
        if (item == null) throw new IllegalArgumentException("item");
        if (isAtCapacity()) throw new RuntimeException("too many items");

        item.setPurchaseDate(LocalDate.now(clock));

        addItem(item);
    }

    private void addItem(Item item) {
        List<Item> existingItems = getItemsNamed(item.getName());
        if (existingItems.isEmpty())
            items.put(item.getName(), existingItems);
        existingItems.add(item);
    }

    private boolean isAtCapacity() {
        return count() == CAPACITY;
    }

    void setClock(Clock clock) {
        this.clock = clock;
    }

    public boolean contains(String name) {
        return items.containsKey(name);
    }

    public List<Item> getItemsNamed(String name) {
        if (!contains(name))
            return new ArrayList<>();
        return items.get(name);
    }

    public int count(String name) {
        return getItemsNamed(name).size();
    }

    public int count() {
        return (int) flatten(items.values()).count();
    }

    public Item getItemNamed(String name) {
        if (!contains(name)) return null;
        return getItemsNamed(name).get(0);
    }

    public List<Item> getItemsExpiringToday() {
        return filterFlattened(items.values(),
                item -> item.getExpirationDate().isEqual(LocalDate.now(clock)));
    }

    public List<Item> getAllExpiredItems() {
        return filterFlattened(items.values(),
                item -> !item.getExpirationDate().isAfter(LocalDate.now(clock)));
    }

    private <T> List<T> filterFlattened(Collection<List<T>> values, Predicate<T> predicate) {
        return flatten(values)
                .filter(predicate)
                .collect(Collectors.toList());
    }

    private <T> Stream<T> flatten(Collection<List<T>> values) {
        return values.stream().flatMap(Collection::stream);
    }
}

File: csharp-pi-pantry/csharp-pi-pantry/Domain/Pantry.cs
using Pipantry.Util;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Pipantry.Domain {
    public class Pantry {
        private IDictionary<string, IList<Item>> items = new Dictionary<string, IList<Item>>();
        private const int CAPACITY = 512;

        public Pantry()
        {
            DateTimeProvider = new DateTimeWrapper();
        }

        public IDateTime DateTimeProvider { get; set; }

        public void Purchase(Item item) {
            if (item == null) throw new ArgumentException("item");
            if (IsAtCapacity()) throw new Exception("too many items");

            item.PurchaseDate = DateTimeProvider.Now;

            Add(item);
        }

        private void Add(Item item) {
            var existingItems = ItemsNamed(item.Name).ToList();
            if (!existingItems.Any())
                items[item.Name] = existingItems;
            existingItems.Add(item);
        }

        private bool IsAtCapacity() {
            return Count() == CAPACITY;
        }

        public bool Contains(string name) {
            return items.ContainsKey(name);
        }

        public IEnumerable<Item> ItemsNamed(string name) {
            if (!Contains(name))
                return new List<Item>();
            return items[name];
        }

        public int Count(string name) {
            return ItemsNamed(name).Count();
        }

        public int Count() {
            return items.Values.SelectMany(i => i).Count();
        }

        public Item ItemNamed(string name) {
            if (!Contains(name)) return null;
            return ItemsNamed(name).FirstOrDefault(); // TODO Item?
        }

        public IEnumerable<Item> getItemsExpiringToday() {
            return items.Values.SelectMany(i => i).Where(item => item.ExpirationDate == DateTimeProvider.Now);
        }

        public IEnumerable<Item> getAllExpiredItems() {
            return items.Values.SelectMany(i => i).Where(item => DateTimeProvider.Now >= item.ExpirationDate);
        }
    }
}
